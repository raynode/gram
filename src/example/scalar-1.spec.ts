import {
  GraphQLBoolean,
  GraphQLInt,
  GraphQLScalarType,
  GraphQLString,
  printSchema,
} from 'graphql'
import {
  createSchemaBuilder,
  FilterCheckFn,
  FilterFn,
  filters,
  isSpecificScalarType,
  Paged,
  Service,
} from '..'

enum AnimalTypes {
  Cat,
  Dog,
}

interface Animal {
  type: AnimalTypes
  name: string
  tame: boolean
  age: number
}

const Animals: Service<Animal> = {
  findOne: async () => null,
  findMany: async () => ({
    page: null,
    nodes: [],
  }),
}

const EatingType = new GraphQLScalarType({
  name: 'EatingType',
  description: 'What is this animal eating',
  serialize: val => val,
})

describe('testing the example 3', () => {
  it('should build the example code', () => {
    const builder = createSchemaBuilder()
    builder.setScalar('EatingType', EatingType)

    const animal = builder.model('Animal', Animals)

    const DateTime = builder.getScalar('DateTime')
    // animal name like 'Fluffy', 'Rex'
    // field is not required
    animal.attr('name', GraphQLString)
    // is it a tame animal
    animal.attr('tame', GraphQLBoolean)
    // age of the animal
    animal.attr('dateOfBirth', DateTime)
    // feeding type of the animal
    animal.attr('feed', EatingType)

    expect(printSchema(builder.build())).toMatchSnapshot()
  })

  it('should handle new filters', () => {
    const builder = createSchemaBuilder()
    builder.setScalar('EatingType', EatingType)

    const checkFn = isSpecificScalarType('EatingType')
    const filterFn = filters.joinFilters([
      filters.equals, // adds equal & not-equal filter
      filters.record, // adds in && not-in filters
    ])
    builder.addFilter(checkFn, filterFn)

    const animal = builder.model('Animal', Animals)

    const DateTime = builder.getScalar('DateTime')
    // animal name like 'Fluffy', 'Rex'
    // field is not required
    animal.attr('name', GraphQLString)
    // is it a tame animal
    animal.attr('tame', GraphQLBoolean)
    // age of the animal
    animal.attr('dateOfBirth', DateTime)
    // feeding type of the animal
    animal.attr('feed', EatingType)

    expect(printSchema(builder.build())).toMatchSnapshot()
  })
})
