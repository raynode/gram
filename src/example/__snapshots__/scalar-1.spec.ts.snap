// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the example 3 should build the example code 1`] = `
"type Animal implements Node {
  name: String
  tame: Boolean
  dateOfBirth: DateTime
  feed: EatingType
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

input AnimalFilter {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input AnimalPage {
  limit: Int
  offset: Int
}

type Animals implements List {
  page: Page
  nodes: [Animal!]!
}

enum AnimalSortOrder {
  name_ASC
  name_DESC
  tame_ASC
  tame_DESC
  dateOfBirth_ASC
  dateOfBirth_DESC
  feed_ASC
  feed_DESC
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
}

input AnimalWhere {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input CreateAnimalData {
  name: String
  tame: Boolean
  dateOfBirth: DateTime
  feed: EatingType
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

scalar DateTime

scalar EatingType

interface List {
  page: Page
  nodes: [Node!]!
}

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Query {
  getAnimal(where: AnimalWhere!, order: AnimalSortOrder): Animal
  getAnimals(order: AnimalSortOrder, page: AnimalPage, where: AnimalWhere!): Animals
}

input UpdateAnimalData {
  name: String
  tame: Boolean
  dateOfBirth: DateTime
  feed: EatingType
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}
"
`;

exports[`testing the example 3 should handle new filters 1`] = `
"type Animal implements Node {
  name: String
  tame: Boolean
  dateOfBirth: DateTime
  feed: EatingType
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

input AnimalFilter {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  feed: EatingType
  feed_not: EatingType
  feed_in: [EatingType!]
  feed_not_in: [EatingType!]
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input AnimalPage {
  limit: Int
  offset: Int
}

type Animals implements List {
  page: Page
  nodes: [Animal!]!
}

enum AnimalSortOrder {
  name_ASC
  name_DESC
  tame_ASC
  tame_DESC
  dateOfBirth_ASC
  dateOfBirth_DESC
  feed_ASC
  feed_DESC
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
}

input AnimalWhere {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  feed: EatingType
  feed_not: EatingType
  feed_in: [EatingType!]
  feed_not_in: [EatingType!]
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input CreateAnimalData {
  name: String
  tame: Boolean
  dateOfBirth: DateTime
  feed: EatingType
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

scalar DateTime

scalar EatingType

interface List {
  page: Page
  nodes: [Node!]!
}

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Query {
  getAnimal(where: AnimalWhere!, order: AnimalSortOrder): Animal
  getAnimals(order: AnimalSortOrder, page: AnimalPage, where: AnimalWhere!): Animals
}

input UpdateAnimalData {
  name: String
  tame: Boolean
  dateOfBirth: DateTime
  feed: EatingType
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}
"
`;
